### ISSUE TITLE
Fix S3247 FP: Rule should not trigger when variable is used for data binding on Razor/MVC views

### ISSUE THREAD

**Issue by cbaerikebc** [SonarSource/sonar-dotnet#8735]
### Description

Rule S3247 now triggers for code on Razor and MVC views, which is great! But it should only do so when the variable is used in a read-only format. If I replace the type cast as SonarQube suggests, but the type cast is part of a data binding scenario, the data binding no longer works.

### Repro steps

ViewModels:
```
public class Base {}

public class Sub : Base {
    public string Property1 { get; set; } = string.Empty;
}
```

View:
```
@model Base

@if (Model is Sub)
{
    <div>
        <input class="form-control" type="text" asp-for="@(((Sub)Model).Property1)" />
        <span asp-validation-for="@(((Sub)Model).Property1)" class="text-danger"></span>
    </div>
}
```

### Expected behavior

Rule does not trigger

### Actual behavior

Rule suggests this code instead, which breaks data binding:
```
@if (Model is Sub sub)
{
    <div>
        <input class="form-control" type="text" asp-for="@(sub.Property1)" />
        <span asp-validation-for="@(sub.Property1)" class="text-danger"></span>
    </div>
}
```

### Known workarounds

None

### Related information

* C#10 in a .NET 6 Razor project
* Visual Studio version 2022
* SonarQube 10.3
* Scan run as part of an Azure DevOps Server 2022 pipeline, using dotnet build

**Comment by cbaerikebc** [comment]
Hi @sebastien-marichal, I'm happy to elaborate.

Data binding in MVC and Razor is done through the `asp-for=""` property. This instructs .NET to display the content of the given variable in the input field when rendering the page. It also writes the content of the input field back to the variable during a POST, so that it can easily be processed by the controller.

For this to work, the variable needs to be part of the Model object, e.g. `@(((Sub)Model).Property1)`

With the suggested code, the input field is instead bound to the `sub` object, which means all those properties in the Model object are empty after a POST.

Long story short: If the variable is used in an `asp-for=""` property, then type casting is needed and you cannot use the `is` statement to re-assign the Model object to another object. Technically, this is only an issue when the data from the input field is sent back to the server - it will display just fine on a read-only page. But that distinction may be hard to detect by SonarQube.

**Comment by martin-strecker-sonarsource** [comment]
The problem is that for `asp-for` the expression `(((Sub)Model).Property1` is bound to a `System.Linq.Expressions.Expression`:
https://github.com/dotnet/aspnetcore/blob/d93b8e28b53f7c5d5cf6722facaef6012efa32b2/src/Mvc/Mvc.ViewFeatures/src/ModelExpression.cs#L8-L11

This is an indication that the code is interpreted instead of executed. We should in general not raise in an expression context. We do have the `IsInExpressionTree` helper method to detect such cases, which (hopefully) works here as well.
https://github.com/SonarSource/sonar-dotnet/blob/8e491ee2944b481970c80991716920ba614e655a/analyzers/src/SonarAnalyzer.Common/Facade/SyntaxFacade.cs#L46

Bottom line: This is a true positive and we should not raise it in an expression context. Other frameworks like EntityFramework may suffer from similar problems.
